sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala

```diff
@@ -0,0 +1,324 @@
+  // ::2:3:4:5:6:7:8  ::2:3:4:5:6:7:8 ::8       ::
+  val ipv6Regex10 = """(fe80:((:([0-9a-fA-F][0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?)?)(:([0-9a-fA-F][0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?)?)?(:([0-9a-fA-F][0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?)?)?(:([0-9a-fA-F][0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?)?)?)?%[\w]+)"""   
+  // fe80::7:8%eth0   fe80::7:8%1     (link-local IPv6 addresses with zone index)
+  val ipv6Regex11 = """(::((ffff|FFFF)(:00?0?0?)?:)?((25[0-5]|(2[0-4]|1?[0-9])?[0-9])\.)((25[0-5]|(2[0-4]|1?[0-9])?[0-9])\.)?((25[0-5]|(2[0-4]|1?[0-9])?[0-9])\.)?(25[0-5]|(2[0-4]|1?[0-9])?[0-9]))""" 
+  // ::255.255.255.255   ::ffff:255.255.255.255  ::ffff:0:255.255.255.255  (IPv4-mapped IPv6 addresses and IPv4-translated addresses)
+  val ipv6Regex12 = """([0-9a-fA-F][0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?:)([0-9a-fA-F][0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?:)?([0-9a-fA-F][0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?:)?([0-9a-fA-F][0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?:)?:((25[0-5]|(2[0-4]|1?[0-9])?[0-9])\.)((25[0-5]|(2[0-4]|1?[0-9])?[0-9])\.)((25[0-5]|(2[0-4]|1?[0-9])?[0-9])\.)(25[0-5]|(2[0-4]|1?[0-9])?[0-9])"""
+  // 2001:db8:3:4::192.0.2.33  64:ff9b::192.0.2.33 (IPv4-Embedded IPv6 Address)
+  val ipv6Regex13 = """(0:0:0:0:0:(0|FFFF|ffff):((25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9]))"""
+  // 0:0:0:0:0:0:13.1.68.3
+  // scalastyle:on
+}
+
+case class GpuParseUrl(children: Seq[Expression], 
+    failOnErrorOverride: Boolean = SQLConf.get.ansiEnabled) 
+  extends GpuExpression with ShimExpression with ExpectsInputTypes {
+
+  def this(children: Seq[Expression]) = this(children, SQLConf.get.ansiEnabled)
+
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[DataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+  override def prettyName: String = "parse_url"
+
+  import GpuParseUrl._
+  
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size > 3 || children.size < 2) {
+      RapidsErrorUtils.parseUrlWrongNumArgs(children.size) match {
+        case res: Some[TypeCheckResult] => return res.get
+        case _ => // error message has been thrown
+      }
+    }
+    super[ExpectsInputTypes].checkInputDataTypes()
+  }
+
+  private def getPattern(key: UTF8String): RegexProgram = {
+    val regex = REGEXPREFIX + key.toString + REGEXSUBFIX
+    new RegexProgram(regex)
+  }
+
+  private def reValid(url: ColumnVector): ColumnVector = {
+    val regex = """([^\s]*\s|([^[]*|[^[]*\[.*].*)%([^0-9a-fA-F]|[0-9a-fA-F][^0-9a-fA-F]|$))"""
```

Comment:
It is used to check if each %?? is a valid escape, but it will take much more time to do another regex. We may want to consider not supporting this check to improve performance.